File: phylo-data-services/app/_scripts/wait_for_development_db.py
#!/usr/bin/env python3
import os
import time

import mariadb  # type: ignore


def check_database():
    """Check the database connection and return True if ready, else False."""
    try:
        connection = mariadb.connect(
            user=os.environ["DB_USER"],
            password=os.environ["DB_PASS"],
            host=os.environ["DB_HOST"],
            port=int(os.environ["DB_PORT"]),
            database=os.environ["DB_DATABASE"],
        )
        if not connection:
            return False

        cursor = connection.cursor()
        cursor.execute("SHOW TABLES;")
        results = cursor.fetchall()
        if len(results) < 1:
            raise Exception("Database is empty!")
        print(f"==> Database tables: {results}")
        print(
            f"==> Database: [{os.environ['DB_DATABASE']}] is ready to receive connection!"
        )
        connection.close()
        return True
    except Exception as e:
        print(f"==> [WARNING] Database check failed: {e}")
        return False


def main():
    retry_time_gap = 5 * 1000  # 5 seconds in milliseconds
    timeout_limit = 60 * 60 * 1000  # 1 hour in milliseconds
    start_time = time.time() * 1000  # Current time in milliseconds
    retry_count = 0

    while not check_database():
        elapsed_time = time.time() * 1000 - start_time
        if elapsed_time > timeout_limit:
            raise Exception(
                f"==> [ERROR] Database is taking too long to connect. Timeout Limit: {timeout_limit} ms!"
            )
        retry_count += 1
        print(
            f"==> Database connection retry: {retry_count} in {retry_time_gap} ms ..."
        )
        time.sleep(retry_time_gap / 1000)


if __name__ == "__main__":
    main()

File: phylomedb6-webapp/_scripts/wait-for-development-db.mjs
import mysql from "mysql2/promise";
import * as dotenv from "dotenv";

if (process.env.NODE_ENV === "development") {
    dotenv.config({ path: "./.env.development" });
} else if (process.env.NODE_ENV === "production") {
    dotenv.config({ path: "./.env.production" });
} else {
    dotenv.config({ path: "./.env.local" });
}

// sleep time expects milliseconds
function sleep(time) {
    return new Promise((resolve) => setTimeout(resolve, time));
}

(async function main() {
    const retryTimeGap = 5 * 1000; // milliseconds (val.: 5 seconds)
    const timeoutLimit = 60 * 60 * 1000; // milliseconds (val.: 1 hour)
    const startTime = new Date();
    let retryCount = 0;
    while ((await checkDatabase()) === false) {
        if (new Date() - startTime > timeoutLimit) {
            throw Error(
                `==> [ERROR] Database is taking too long to connect. Timeout Limit: ${timeoutLimit}!`,
            );
        }
        retryCount++;

        await sleep(retryTimeGap);
    }
})();

async function checkDatabase() {
    try {
        const connection = await mysql.createConnection({
            user: process.env["DB_USER"],
            password: process.env["DB_PASS"],
            host: process.env["DB_HOST"],
            port: process.env["DB_PORT"],
            database: process.env["DB_DATABASE"],
        });
        const [results, _fields] = await connection.query("show tables;");
        if (results.length < 1) {
            throw Error("phylomedb6 database is empty!");
        }
        connection.end();
        return true;
    } catch (err) {
        return false;
    }
}
